src/mscorlib/shared/System/IO/FileStream.cs

```diff
@@ -81,8 +81,31 @@ public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferS
 
         [Obsolete("This constructor has been deprecated.  Please use new FileStream(SafeFileHandle handle, FileAccess access, int bufferSize, bool isAsync) instead, and optionally make a new SafeFileHandle with ownsHandle=false if needed.  http://go.microsoft.com/fwlink/?linkid=14202")]
         public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferSize, bool isAsync)
-            : this(new SafeFileHandle(handle, ownsHandle), access, bufferSize, isAsync)
         {
+            SafeFileHandle safeHandle = new SafeFileHandle(handle, ownsHandle: ownsHandle);
+            try
+            {
+                ValidateAndInitFromHandle(safeHandle, access, bufferSize, isAsync);
+            }
+            catch
+            {
+                // We don't want to take ownership of closing passed in handles
+                // *unless* the constructor completes successfully.
+                GC.SuppressFinalize(safeHandle);
+
+                // This would also prevent Close from being called, but is unnecessary
+                // as we've removed the object from the finalizer queue.
+                //
+                // safeHandle.SetHandleAsInvalid();
```

Comment:
> The assertion this change is making is that we don't own the handle if we haven't finished constructing. Dispose() won't close the handle if it is set as invalid.

I'm not convinced that's valid for the IntPtr ctor.  The calling code is handing off a raw IntPtr to the FileStream saying "it's now your responsibility to close this, nothing else will".

That's very different from the ctors that take a SafeHandle, where the SafeHandle itself knows whether it owns the IntPtr, and where worst case it'll have its finalizer handle the cleanup.